Abstract entity types as intersection of types

roshandathathri opened the pull request:
Entities can be nodes or edges

Rename TypeSetID as EntityTypeID

Add more functions to the API for entity types

ddn0 commented on tools/cypher-compile/CMakeLists.txt:
```diff
@@ -5,5 +5,3 @@ install(TARGETS cypher-compile
   COMPONENT tools
 )
-
-add_subdirectory(test)
```
cypher-compile looks like a test as well or is it intended to be the full cypher interpreter?

l-hoang replied:
It's not really a test more than it is a tool to just run

l-hoang replied:
More context: you can't run cypher-compile without specifying some user query.

I use it to get an idea of what the parse tree looks like from libcypher-parser as well when I'm adding new features to the compiler.

ddn0 replied:
FWIW, seems like some portion of this functionality might be suitable as a unit test of the parser. I can't imagine that users would want to see parse trees directly :D 

I'm fine with leaving cypher-compiler as it is now because we don't yet have those unit tests. But it seems like its future also belongs in `libquery/test` IMHO.

l-hoang replied:
When you put it that way, yes, cypher-compile isn't a user tool more than it is a development tool.

roshandathathri committed:
Abstract entity types as intersection of types

Entities can be nodes or edges

Rename TypeSetID as EntityTypeID

Add more functions to the API for entity types

arthurp commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -155,23 +155,30 @@ class KATANA_EXPORT GraphTopology {
+  /// This type may either be an atomic type or an intersection of atomic types
+  /// EntityTypeID is represented using 8 bits
+  using EntityTypeID = uint8_t;
+  static constexpr EntityTypeID kUnknownType = EntityTypeID{0};
```
Does `kUnknownType` mean that the value could have any type? If so then it's the union of all atomic types. This is usually called `Top`.

roshandathathri replied:
No, it means the type is unspecified/unknown. It cannot be any type.

arthurp replied:
If we do not know the  type of some entity, we can still say that it must have SOME type. And if it has some type, then it is the member of the union of all types. And that union is called "top" or sometimes "any".

roshandathathri replied:
No, it is not equivalent to "any" type. If a user requests nodes with a particular type, this node with unknown type will not be returned. It is a special type. We can enforce that all nodes/edges/entities must have types but we are not planning to do that (at least in the near future).

arthurp replied:
So you view Unknown as it's own type which is a subtype of Any and without any relationship to any other type? That's reasonable I guess. In most systems, you wouldn't have a name for this separate from Any. Instead, values which have the most specific type Any are implicitly unknown because you cannot determine a more specific type for them.

The only issue I can see with having an Unknown type is that since types are not disjoint, having type Unknown doesn't preclude having other types. Encoding unknown by having Any as the most specific type of a value avoids this issue since having any other type implicitly means we know something about the value so Any isn't the most specific type any more.

arthurp commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -155,23 +155,30 @@ class KATANA_EXPORT GraphTopology {
+  using EntityTypeID = uint8_t;
+  static constexpr EntityTypeID kUnknownType = EntityTypeID{0};
+  static constexpr EntityTypeID kInvalidType =
+      std::numeric_limits<EntityTypeID>::max();
```
`numeric_limits` is unsafe. For instance if EntityTypeID were made into an OpaqueID then this would return 0. Use `boost::math::tools::max_value` it will fail if there is no maximum instead of producing terrible bugs.

roshandathathri replied:
This is the sentinel value, right? Can't we do this when we move this to OpaqueIDs?

arthurp replied:
Yes. OpaqueIDs provide a sentinal, so you could definitely use that here if we change EntityTypeID into an OpaqueID (which I would love to see).

arthurp commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -155,23 +155,30 @@ class KATANA_EXPORT GraphTopology {
+  using EntityTypeIDToSetOfTypeNamesMap = std::vector<SetOfTypeNames>;
+  /// A map from the type name to the set of the EntityTypeIDs that intersect it
+  using TypeNameToSetOfEntityTypeIDsMap =
+      std::unordered_map<std::string, SetOfEntityTypeIDs>;
```
You seem to be adding "atomic" to specify which types you are naming in some places, but not others. Is there a reason?

roshandathathri replied:
Yes, some of these are only applicable for atomic types.

arthurp commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -395,11 +409,11 @@ class KATANA_EXPORT PropertyGraph {
+  /// Construct node & edge EntityTypeIDs from node & edge properties
+  /// Also constructs metadata to convert between atomic types and EntityTypeIDs
+  /// Assumes all boolean or uint8 properties are atomic types
   /// TODO(roshan) move this to be a part of Make()
```
I think this TODO is fairly urgent, BTW. Is there a reason not to just add the call and make this private?

roshandathathri replied:
I know. It needs to be moved to the partitioner. It's not as simple as moving it into the constructor.

arthurp replied:
Is there a JIRA issue for this? I think it's worth tracking.

arthurp commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -473,80 +487,200 @@ class KATANA_EXPORT PropertyGraph {
-  size_t GetNodeTypesNum() const {
-    return node_type_name_to_type_set_ids_.size();
+  /// \returns the number of node atomic types
+  size_t GetNodeAtomicTypesNum() const {
```
Since we are renaming: Having "Num" at the end reads weirdly. Could we have `GetNodeAtomicTypesCount` or `GetNumNodeAtomicTypes`?

arthurp replied:
Same for several methods.

roshandathathri replied:
That was the norm in this class. I don't mind changing it but it needs to be consistent. EIther `GetNum<x>` or `Num<x>` is fine with me.

amberhassaan replied:
Agree with Arthur.

arthurp commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -473,80 +487,200 @@ class KATANA_EXPORT PropertyGraph {
+  const SetOfEntityTypeIDs& EdgeTypeNameToEntityTypeIDs(
+      const std::string& name) const {
+    return edge_type_name_to_entity_type_ids_.at(name);
+  }
```
I think these should be removed. If there is a need to get all supertypes of a given type, there can be a `GetSupertypes` method to do that taking an `EntityTypeID`. That separates name look ups from type relationships.

roshandathathri replied:
No, they are required. They are used by querying to do set operations like intersection/union/difference on the super-types. 

I like the super-types. I'll change the function name to `NodeTypeNameToSupertypes`

arthurp replied:
Why does the partitioner need to go directly from name to supertypes? We had talked about separating ID look up from supertype look up. So the query can do `GetSupertypes(GetEntityTypeID(name))`.

roshandathathri replied:
Querying, not partitioner (maybe that's what you meant). 

I agree that the function call `GetSupertypes(GetEntityTypeID(name))` is better.  The datatype returned by `GetSupertypes` needs to be a bitset because querying does set operations like intersection/union/difference on the super-types.

I'll make another pass over it. I understand your requirements better now.

arthurp replied:
I did mean querying. I just did a thinko. :-)

As long as the bitset can be viewed and operated on exactly like a set of EntityTypeIDs. I'm fine with that. A bitset representation makes sense. I only care about the API it provides.

roshandathathri replied:
Great, then I'll drop the "range" functions.

If I recall, I assumed that you didn't want the representation of a "set" to creep into the API. In this case, it would. That's fine, right?

arthurp replied:
Sure. Smaller API is generally better. So I'm in favor of dropping methods. :-)

arthurp commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -473,80 +487,200 @@ class KATANA_EXPORT PropertyGraph {
+        EdgeEntityTypeIDToAtomicSubtypeNames(edge_entity_type_id);
+    ;
+    return MakeStandardRange(type_names.cbegin(), type_names.cend());
+  }
```
Shouldn't these return ranges of `EntityTypeID` not string? Also why include "range" in the name? If they are just wrappers around `EdgeEntityTypeIDToAtomicSubtypeNames` I don't see their usefulness at all. You can always just iterate the `SetOfTypeNames`.

roshandathathri replied:
I misunderstood your API requirements. 

Returning a range of entity-type-ids requires materializing a vector. I could do that but it's not used currently and when it is used, there will be overheads.

amberhassaan replied:
I'm OK with returning an iterator range, but I'd like you to remove the variants that return Sets (above these 2 methods). Those shouldn't be needed given that these 2 methods return ranges.

arthurp replied:
I don't see why you would need to materialize the vector to return a range. The range could be a filter or map operation over the internal data structures of `PropertyGraph`.

roshandathathri replied:
I agree with Arthur that there should be functions to convert atomic-name and `EntityTypeID` but all other functions should be in terms of `EntityTypeID`. Returning a set of `EntityTypeID` is required to do set operations - range won't suffice for that. Returning a range requires materialization because `std::bitset` does not provide an iterator for set bits. Arthur, I did not understand what you meant by filter or map?

arthurp replied:
By filter or map, I just meant we could create a range/iterators that computes the values needed during iteration, eliminating the need to materialize. I said filter/map because that's how you express these kinds of lazy collection operations in many languages. C++ doesn't actually have such operations, but the concepts still exist and can be implemented.

Just because std::bitset doesn't provide such an iterator doesn't mean we cannot.

One thing we could do is have a special collection class that uses a bitset representation internally and provides the required iterators and a way to materialize the set or bitset when required.

roshandathathri replied:
I think I now understand. I had implemented such an iterator for `DynamicBitset`. Maybe we can repurpose that iterator for this.

arthurp replied:
All this said. I'm gonna back off on this. I don't want to create too much work without much gain. C++ makes good abstraction so difficult that it may not be worth it. Especially given that the idiomatic style of C++ tends to be very poorly abstracted.

roshandathathri replied:
Thanks both of you for all the comments/suggestions. I'll handle this soon.

arthurp reviewed (changes requested):
There seems to be a mix of names and ID used in the API. I would rather consistently use IDs everywhere except the lookup function which translates the name to an ID. That way there is no confusion as to what type is needed.

Also let's try to separate type relationships into explicitly named functions. Things like `GetAllSubtypes` or `GetAllSuperTypes`. Having these operations mixed in with the lookups is confusing.

roshandathathri commented:
> There seems to be a mix of names and ID used in the API. I would rather consistently use IDs everywhere except the lookup function which translates the name to an ID. That way there is no confusion as to what type is needed.
> 
> Also let's try to separate type relationships into explicitly named functions. Things like `GetAllSubtypes` or `GetAllSuperTypes`. Having these operations mixed in with the lookups is confusing.

Makes sense. Will do.

amberhassaan commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -155,23 +155,30 @@ class KATANA_EXPORT GraphTopology {
-  /// A map from the type name to the set of the TypeSetIDs that contain it
-  using TypeNameToSetOfTypeSetIDsMap =
-      std::unordered_map<std::string, SetOfTypeSetIDs>;
+  using TypeNameIterator = std::unordered_set<std::string>::const_iterator;
```
Use the alias above. i.e. `SetofTypeNames::const_iterator`

amberhassaan commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -473,80 +487,200 @@ class KATANA_EXPORT PropertyGraph {
+  /// NB: even if it exists, it may not be the most specific type for any node
+  /// (returns true for kUnknownType)
+  bool HasNodeEntityType(EntityTypeID node_entity_type_id) const {
+    return node_entity_type_id < node_entity_type_id_to_type_names_.size();
```
Shouldn't there be a stronger check here, e.g., `&& !node_entity_type_id_to_type_names_[id].empty()` ?

amberhassaan commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -473,80 +487,200 @@ class KATANA_EXPORT PropertyGraph {
+  /// \returns true iff the node type @param sub_type is a
+  /// sub-type of the node type @param super_type
+  /// (assumes that the sub_type and super_type EntityTypeIDs exists)
+  bool IsNodeSubtypeOf(EntityTypeID sub_type, EntityTypeID super_type) const {
```
This is a bad implementation. We're checking for strings from one set in another. This is where we should be keeping maps to bit-sets (instead of strings) and compute the result much cheaper.

amberhassaan reviewed (approved):
Overall, this is good. Please point me (us) to the enterprise PR where some of this API will be used (querying I suppose?).

roshandathathri commented:
> These names are confusing. (there's a general naming issue due to nodes and edges). Should it be `NodeHasSubType` and `EdgeHasSubtype`

Which functions are you referring to? I think `GetNodeType` and `HasNodeType` are appropriate.

arthurp commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -473,80 +495,172 @@ class KATANA_EXPORT PropertyGraph {
+    return edge_entity_type_id_to_atomic_entity_type_ids_.size();
+  }
+
+  /// \returns true iff a node atomic type @param name exists
```
nit: For the future `@param` produces a whole new block. To reference a param you want `\p` (https://www.doxygen.nl/manual/commands.html#cmdp).

arthurp commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -473,80 +495,172 @@ class KATANA_EXPORT PropertyGraph {
+  }
+
+  /// \returns true iff a node atomic type @param name exists
+  /// NB: no node may have a type that intersects with this atomic type
```
You could just say "no node may have this atomic type". It means the same thing.

arthurp reviewed (approved):
LGTM. There are a few remaining comments, but they can/should all wait until another PR.

arthurp commented:
> > These names are confusing. (there's a general naming issue due to nodes and edges). Should it be `NodeHasSubType` and `EdgeHasSubtype`
> 
> Which functions are you referring to? I think `GetNodeType` and `HasNodeType` are appropriate.

I also thought it was slightly confusing to have `HasNodeType` because it's not clear if you mean that the graph has the node type or if a specific node does. That might be the confusion. I don't know how to fix it though.

roshandathathri commented:
> > > These names are confusing. (there's a general naming issue due to nodes and edges). Should it be `NodeHasSubType` and `EdgeHasSubtype`
> > 
> > 
> > Which functions are you referring to? I think `GetNodeType` and `HasNodeType` are appropriate.
> 
> I also thought it was slightly confusing to have `HasNodeType` because it's not clear if you mean that the graph has the node type or if a specific node does. That might be the confusion. I don't know how to fix it though.

Thanks. Will change it to `DoesNodeHaveType` and `GetTypeOfNode`

roshandathathri commented:
> Please point me (us) to the enterprise PR where some of this API will be used (querying I suppose?).

https://github.com/KatanaGraph/katana-enterprise/pull/1568

e-mcginnis reviewed (approved):
LGTM

GioLomia reviewed (approved):
Looks good.
